docs/decisions/0015-outbox-pattern-and-production-modes.rst

```diff
@@ -0,0 +1,95 @@
+- **On-commit send**: The event is only sent from a ``django.db.transaction.on_commit`` callback. This preserves atomicity in the failure case, but the IDA could crash after transaction commit but before calling the broker -- or more commonly, the broker could be down or unreachable, and all of the on-commit-produced events would be lost during that interval. Ordering is also not preserved here.
+
+We currently use an ad-hoc mix of immediate and on-commit send in edx-platform, depending on how particular OpenEdxPublicSignals are emitted. For example, the code path for ``COURSE_CATALOG_INFO_CHANGED`` involves an explicit call to ``django.db.transaction.on_commit`` in order to ensure an on-commit send is used. But most signals do not have any such call, and are likely sent immediately. This uncontrolled state reflects our iterative approach to the event bus as well as our choice to start with events that are backed by other synchronization measures which can correct for consistency issues. However, we'd like to start handling events that require stronger reliability guarantees, such as those in the ecommerce space.
+
+Decision
+********
+
+We will implement the transactional outbox pattern (or just "outbox pattern") in order to allow binding event production to database transactions. Events will default to on-commit send, but openedx-events configuration will be enhanced to allow configuring each event to a choice of production mode (on-commit or outbox).
+
+In the outbox pattern, events are not produced as part of the request/response cycle, but are instead appended to an "outbox" database table within the transaction. A worker process operating in a separate transaction works through the list in order, producing them to the message broker and removing them once the broker has acknowledged them. This is the standard solution to the dual-write problem and is likely the only way to meet all of the criteria. Atomicity is ensured by bringing the *intent* to send an event into the transaction's ACID guarantees. Transaction commits also impose a meaningful ordering across all hosts using the same database.
+
+openedx-events will change to support two producer modes for sending events when ``send(...)`` is called:
+
+- ``on-commit``: Delay sending to the event bus until after the current transaction commits, or immediately if there is no open transaction (as might occur in a worker process).
+
+  This requires ensuring that any events that are currently being explicitly sent on-commit are changed to call ``get_producer().send(...)`` directly, after appropriate per-event configuration. ``emit_catalog_info_changed_signal`` is a known example of this.
```

Comment:
Here's where we currently delay the sending of a specific event until after the transaction commits: https://github.com/openedx/edx-platform/blob/a74f510f71c19142aeb380facdc2cdfecf16c38a/cms/djangoapps/contentstore/signals/handlers.py#L153 (Note that it currently has an additional race condition, in that the data for the event is gathered _after_ the transaction has committed.)

We'd be able to replace this `transaction.on_commit` call with something like `COURSE_CATALOG_INFO_CHANGED.send_event(*_create_catalog_data_for_signal(course_key))` (computing the event body _inside_ the transaction). The default configuration of `COURSE_CATALOG_INFO_CHANGED` would be to send it in `on-commit` mode, largely preserving the existing behavior -- but without that race condition.

Reply:
I think this raises some important caveats that are worth documenting, maybe in "consequences". django-jaiminho stores the args/kwargs and only executes the function call to send the message after (maybe quite a while after) the transaction is committed, so it is important to be very mindful of both the size of what you're potentially sending in and the fact that you're not actually acting in the same transaction as the rest of the code, which may have some unintended consequences.

Things like passing in a lazy Django queryset can cause it to get evaluated, so if you're planning on doing further filtering in the sending function, that should get bubbled up to the caller. There's also a 65k limit to the `dill`'d stored args, so there's a chance of inadvertently overrunning that with an evaluated queryset and either losing the event due to error or saving an unserializable blob to the database.

Similarly we'd need to be on guard against pulling any new data in the sending function since it could be operating on data hours later than when it was queued, breaking the ordering guarantees. 

With `dill` in the picture folks will also need to think about the usual `pickle` security problems, and potential serialization issues that they wouldn't have otherwise had to.

Reply:
@timmc-edx: can we add to the ADR the example you [just mentioned](https://github.com/openedx/openedx-events/pull/292/files#r1423141734) about the replacement? 

Also, in these lines you mention:
```suggestion
  This requires ensuring that any events that are currently being explicitly sent on-commit are changed to call ``get_producer().send(...)`` directly, after appropriate per-event configuration. ``emit_catalog_info_changed_signal`` is a known example of this.
```
But the example is different. Should the ADR be updated with the proper example?

Reply:
@bmtcril Yes, I think what we want to do here is only use the jaiminho decorator on a very simple, very stable function that only accepts pre-serialized data -- maybe JSON strings. I don't think jaiminho's overall approach in that regard is very safe, but it looks like there's a safe way to use it. (We could also just call `jaiminho.publish_strategies.publish` directly.)

Reply:
@mariajgrimaldi Sorry, yes, I was still thinking of the older version of openedx-events where we explicitly call send on the producer. Now that we have configuration-based event publishing, everything is handled by openedx-events' signal handlers.

Specifically, [`openedx_events.apps.general_signal_handler`](https://github.com/openedx/openedx-events/blob/1a5f9ca8ff03a573bdd44a8abee9b77aba26d80e/openedx_events/apps.py#L13C17-L13C17) is currently the receiver for OpenEdxPublicSignals, and it makes a call to `get_producer().send(...)`. `general_signal_handler` would change to look at the publishing mode for the event and either store the event in the outbox or schedule the `get_producer().send(...)` to be run on-commit.

If that sounds right to you, I can update the Implementation Plan section.

Reply:
@timmc-edx: thank you! That would be ideal.

Reply:
@mariajgrimaldi Should be taken care of by the recent commits -- let me know what you think.

Reply:
@timmc-edx: Thanks again. Now it's crystal clear.